### PR TITLE
fix(sec): avoid sql injection in virtual metric form

### PR DESCRIPTION
**Pull request by kduret**
## Description

avoid sql injection in virtual metric form

**Fixes** MON-12446

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [x] 20.10.x
- [x] 21.04.x
- [x] 21.10.x
- [x] 22.04.x (master)

### CODE REVIEW

**Review by callapa — CHANGES_REQUESTED** [review]

File: www/include/views/virtualMetrics/DB-Func.php [hunk 1]
```diff
@@ -229,133 +228,231 @@ function insertVirtualMetricInDB()
...
+
     global $form, $pearDB, $pearDBO;
-    $ret = array();
+
     $ret = $form->getSubmitValues();
-    $rq = "UPDATE virtual_metrics ";
-    $rq .= "SET `index_id` = ";
-    if (isset($ret["host_id"]) && preg_match('/\d+\-\d+/', $ret["host_id"])) {
+
+    $indexId = isset($ret["host_id"]) && $ret["host_id"] !== null
```
[the remark this answers is not in the document]
Comment (kduret) [reply www/include/views/virtualMetrics/DB-Func.php:335]: getIndexIdFromHostServiceId expect a string, so we should check if it's not null

File: www/include/views/virtualMetrics/DB-Func.php [hunk 2]
```diff
@@ -229,133 +228,231 @@ function insertVirtualMetricInDB()
...
+
+    $updateStatement->bindValue(
+        ':hidden',
+        $ret['vhidden'] ?? null,
+        \PDO::PARAM_STR
+    );
+
+    $updateStatement->bindValue(
+        ':comment',
+        isset($ret["comment"]) && $ret["comment"] != null
```
[the remark this answers is not in the document]
Comment (kduret) [reply www/include/views/virtualMetrics/DB-Func.php:407]: htmlentities expects a string, so we should check if it's not null

File: www/include/views/virtualMetrics/DB-Func.php [hunk 3]
```diff
@@ -229,133 +228,231 @@ function insertVirtualMetricInDB()
...
+
+    $updateStatement->bindValue(
+        ':index_id',
+        $indexId,
+        \PDO::PARAM_INT
+    );
+
+    $updateStatement->bindValue(
+        ':vmetric_name',
+        isset($ret["vmetric_name"]) && $ret["vmetric_name"] != null
```
[the remark this answers is not in the document]
Comment (kduret) [reply www/include/views/virtualMetrics/DB-Func.php:363]: htmlentities expects a string, so we should check if it's not null

File: www/include/views/virtualMetrics/DB-Func.php [hunk 4]
```diff
@@ -229,133 +228,231 @@ function insertVirtualMetricInDB()
...
+
+    $insertStatement->bindValue(
+        ':hidden',
+        $ret['vhidden'] ?? null,
+        \PDO::PARAM_STR
+    );
+
+    $insertStatement->bindValue(
+        ':comment',
+        isset($ret["comment"]) && $ret["comment"] != null
```
[the remark this answers is not in the document]
Comment (kduret) [reply www/include/views/virtualMetrics/DB-Func.php:304]: htmlentities expects a string, so we should check if it's not null

File: www/include/views/virtualMetrics/DB-Func.php [hunk 5]
```diff
@@ -229,133 +228,231 @@ function insertVirtualMetricInDB()
...
+
+    $insertStatement->bindValue(
+        ':index_id',
+        $indexId,
+        \PDO::PARAM_INT
+    );
+
+    $insertStatement->bindValue(
+        ':vmetric_name',
+        isset($ret["vmetric_name"]) && $ret["vmetric_name"] != null
```
[the remark this answers is not in the document]
Comment (kduret) [reply www/include/views/virtualMetrics/DB-Func.php:260]: htmlentities expects a string, so we should check if it's not null

File: www/include/views/virtualMetrics/DB-Func.php [hunk 6]
```diff
@@ -229,133 +228,231 @@ function insertVirtualMetricInDB()
...
+    global $form, $pearDB, $pearDBO;
+
     $ret = $form->getSubmitValues();
 
-    $rq = "INSERT INTO `virtual_metrics` ( `vmetric_id` , `index_id`, `vmetric_name`, `def_type` , `rpn_function`, " .
-        "`unit_name` , `warn`, `crit`, `hidden` , `comment` , `vmetric_activate`, `ck_state`) ";
-    $rq .= "VALUES ( NULL, ";
+    $indexId = isset($ret["host_id"]) && $ret["host_id"] !== null
+        ? getIndexIdFromHostServiceId($pearDBO, $ret["host_id"])
+        : null;
```
[the remark this answers is not in the document]
Comment (kduret) [reply www/include/views/virtualMetrics/DB-Func.php:241]: getIndexIdFromHostServiceId expects a string, so we should check if it's not null

**Review by callapa — APPROVED** [review]

**Review by jeremyjaouen — APPROVED** [review]

**Pull request merged**